File: Category/Binary Search/Easy/69/69.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x <= 1:
            return x
        l = 1
        h = x
        while l <= h:
            m = l + (h - l) // 2
            sqrt = x // m
            if m == sqrt:
                return sqrt
            elif m > sqrt:
                h = m - 1
            else:
                l = m + 1
        return h

File: Category/Binary Search/Easy/69/69.java
class Solution {
    public int mySqrt(int x) {
        int i = 0;
        // j = x 而不是 x + 1因为容易有越界问题，当x = 2^31 -1时j = x + 1会变成0
        int j = x;
        while (i <= j) {
            int mid = i + (j - i) / 2;
            if ((long) mid * mid == x) {
                return mid;
            } else if ((long) mid * mid < x) {
                i = mid + 1;
            } else {
                j = mid - 1;
            }
        }
        return i - 1; 
    }
}